src/Interpreters/Cache/QueryResultCache.cpp

```diff
@@ -0,0 +1,343 @@
+{
+    SipHash hash;
+    hash.update(key.ast->getTreeHash());
+    hash.update(key.partition_key);
+    auto res = hash.get64();
+    return res;
+}
+
+namespace
+{
+
+auto is_stale = [](const QueryResultCache::Key & key)
+{
+    return (key.expires_at < std::chrono::system_clock::now());
+};
+
+}
+
+QueryResultCache::Writer::Writer(std::mutex & mutex_, Cache & cache_, const Key & key_,
+    size_t & cache_size_in_bytes_, size_t max_cache_size_in_bytes_,
+    size_t max_cache_entries_,
+    size_t max_entry_size_in_bytes_, size_t max_entry_size_in_rows_,
+    std::chrono::milliseconds min_query_runtime_)
+    : mutex(mutex_)
+    , cache(cache_)
+    , key(key_)
+    , cache_size_in_bytes(cache_size_in_bytes_)
+    , max_cache_size_in_bytes(max_cache_size_in_bytes_)
+    , max_cache_entries(max_cache_entries_)
+    , max_entry_size_in_bytes(max_entry_size_in_bytes_)
+    , max_entry_size_in_rows(max_entry_size_in_rows_)
+    , min_query_runtime(min_query_runtime_)
+{
+    if (auto it = cache.find(key); it != cache.end() && !is_stale(it->first))
+        skip_insert = true; /// Key already contained in cache and did not expire yet --> don't replace it
+}
+
+void QueryResultCache::Writer::finalizeWrite()
+{
+    if (skip_insert)
+        return;
+
+    if (std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - query_start_time) < min_query_runtime)
+        return;
+
+    auto to_single_chunk = [](const Chunks & chunks_) -> Chunk
+    {
+        if (chunks_.empty())
+            return {};
+
+        Chunk res = chunks_[0].clone();
+        for (size_t i = 1; i != chunks_.size(); ++i)
+            res.append(chunks_[i]);
+        return res;
+    };
+
+    auto query_result = to_single_chunk(partial_query_results);
+    new_entry_size_in_bytes = query_result.allocatedBytes(); // updated because compression potentially affects the size of the single chunk vs the aggregate size of individual chunks
+
+    std::lock_guard lock(mutex);
+
+    if (auto it = cache.find(key); it != cache.end() && !is_stale(it->first))
+        return; /// same check as in ctor because a parallel Writer could have inserted the current key in the meantime
+
+    auto sufficient_space_in_cache = [this]() TSA_REQUIRES(mutex)
+    {
+        return (cache_size_in_bytes + new_entry_size_in_bytes <= max_cache_size_in_bytes) && (cache.size() + 1 <= max_cache_entries);
+    };
+
+    if (!sufficient_space_in_cache())
+    {
+        size_t removed_items = 0;
+        /// Remove stale entries
+        for (auto it = cache.begin(); it != cache.end();)
+            if (is_stale(it->first))
+            {
+                cache_size_in_bytes -= it->second.allocatedBytes();
+                it = cache.erase(it);
+                ++removed_items;
+            }
+            else
+                ++it;
+        LOG_DEBUG(&Poco::Logger::get("QueryResultCache"), "Removed {} stale entries", removed_items);
```

Comment:
someone will soon ask you to change to `LOG_TRACE` )

Reply:
Good idea, done.